hsds/servicenode_lib.py

```diff
@@ -341,6 +342,158 @@ async def getDsetJson(app, dset_id,
+
+    if h5path and tgt_id:
+        msg = "putLink - provide tgt_id or h5path, but not both"
+        log.warn(msg)
+        raise HTTPBadRequest(reason=msg)
+
+    link_json = {}
+
+    if tgt_id:
+        if not isValidUuid(tgt_id):
+            msg = f"putLink with invalid id: {tgt_id}"
+            log.warn(msg)
+            raise HTTPBadRequest(reason=msg)
+        link_json["id"] = tgt_id
+        link_class = "H5L_TYPE_HARD"
+    elif h5path:
+        link_json["h5path"] = h5path
+        # could be hard or soft link
+        if h5domain:
+            link_json["h5domain"] = h5domain
+            link_class = "H5L_TYPE_EXTERNAL"
+        else:
+            # soft link
+            link_class = "H5L_TYPE_SOFT"
+    else:
+        msg = "PUT Link with no id or h5path keys"
+        log.warn(msg)
+        raise HTTPBadRequest(reason=msg)
+
+    link_json["class"] = link_class
+
+    # for hard links, verify that the referenced id exists and is in
+    # this domain
+    if link_class == "H5L_TYPE_HARD":
+        tgt_id = link_json["id"]
+        ref_json = await getObjectJson(app, tgt_id, bucket=bucket)
+        group_json = await getObjectJson(app, group_id, bucket=bucket)
+        if ref_json["root"] != group_json["root"]:
+            msg = "Hard link must reference an object in the same domain"
+            log.warn(msg)
+            raise HTTPBadRequest(reason=msg)
+
+    # ready to add link now
+    req = getDataNodeUrl(app, group_id)
+    req += "/groups/" + group_id + "/links"
+    log.debug(f"PUT links - PUT request: {req}")
+    params = {"bucket": bucket}
+
+    data = {"links": {title: link_json}}
+
+    put_rsp = await http_put(app, req, data=data, params=params)
+    log.debug(f"PUT Link resp: {put_rsp}")
+    if "status" in put_rsp:
+        status = put_rsp["status"]
+    else:
+        status = 201
+    return status
+
+
+async def putHardLink(app, group_id, title, tgt_id=None, bucket=None):
+    """ create a new hard link.  Return 201 if this is a new link,
+      or 201 if it's a duplicate of an existing link """
+
+    status = await putLink(app, group_id, title, tgt_id=tgt_id, bucket=bucket)
+    return status
+
+
+async def putSoftLink(app, group_id, title, h5path=None, bucket=None):
+    """ create a new hard link.  Return 201 if this is a new link,
+      or 201 if it's a duplicate of an existing link """
+
+    status = await putLink(app, group_id, title, h5path=h5path, bucket=bucket)
+    return status
+
+
+async def putExternalLink(app, group_id, title, h5path=None, h5domain=None, bucket=None):
+    """ create a new hard link.  Return 201 if this is a new link,
+      or 201 if it's a duplicate of an existing link """
+
+    status = await putLink(app, group_id, title, h5path=h5path, h5domain=h5domain, bucket=bucket)
+    return status
+
+
+async def deleteLinks(app, group_id, titles=None, separator="/", bucket=None):
+    """ delete the requested set of attributes from the given object """
```

Comment:
`attributes from the given object` -> `links from the given group`

Reply:
Fixed